Move Reaction.create delay calls to ActiveJob (#3136)

johncurcio opened the pull request:
<!--- Prepend PR title with [WIP] if work in progress. Remove when ready for review. -->

<!--- For a timely review/response, please avoid force-pushing additional commits if your PR already received reviews or comments -->

## What type of PR is this? (check all applicable)

- [x] Refactor

## Description

Move ``Reaction.delay.create`` calls to active job ``perform_later`` calls.

## Related Tickets & Documents

#3136

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)

no UI changes

## Added to documentation?

- [x] no documentation needed

## [optional] What gif best describes this PR or how it makes you feel?

![alt_text](gif_link)

rhymes commented on app/jobs/reactions/reaction_create_job.rb:
```diff
@@ -0,0 +1,14 @@
+module Reactions
+  class ReactionCreateJob < ApplicationJob
```
This can just be `Reactions::CreateJob`

rhymes commented on app/jobs/reactions/reaction_create_job.rb:
```diff
@@ -0,0 +1,14 @@
+    queue_as :reaction_create
+
+    def perform(user_id:, reactable_id:, reactable_type:, category:)
+      Reaction.create(
```
I think this job should load both the user and the reactable from the DB and exit if any of them is not available.

You can check `app/jobs/articles/update_analytics_job.rb` ad an example.

This way the job will just exit if in the meantime those resources have disappeared.

johncurcio replied:
@rhymes there's something I don't understand about the reactable. From what I've looked up, I can get a reactable like this:

```ruby
reaction = Reaction.find_by(reactable_id: reactable_id, reactable_type: reactable_type)
reactable = reaction&.reactable
```

My job is essentially creating a new ``Reaction``. If I load the reactable this way, then to create a new ``Reaction`` using my job there must be another ``Reaction`` in the DB. That does not make sense to me. Am I missing something?

rhymes replied:
You do have a point, I was thinking of loading the reactable without loading the reaction.

Something like `reactable_type.constantize.find(reactable is)` IIRC

rhymes commented on app/controllers/api/v0/reactions_controller.rb:
```diff
@@ -24,7 +24,7 @@ def onboarding
         reactable_ids = JSON.parse(params[:articles]).map { |article| article["id"] }
         reactable_ids.each do |article_id|
-          Reaction.delay.create(
+          ReactionCreateJob.perform_later(
```
Shouldn't this be namespaced?

johncurcio replied:
You're right. I'm going to fix this.

rhymes commented on spec/jobs/reactions/reaction_create_job_spec.rb:
```diff
@@ -0,0 +1,6 @@
+require "jobs/shared_examples/enqueues_job"
+
+RSpec.describe Reactions::ReactionCreateJob, type: :job do
+  include_examples "#enqueues_job", "reaction_create", [{ user_id: 786, reactable_id: 790, reactable_type: "Article", category: "readinglist" }]
```
Keep in mind that the tests should also test the content of the job, take a look at `spec/jobs/messages/send_push_job_spec.rb`

johncurcio commented:
@rhymes I added some testing to the content of the job and fixed the other concerns you had. If you think I can improve the tests, please let me know.

lightalloy commented on app/jobs/reactions/create_job.rb:
```diff
@@ -0,0 +1,19 @@
+
+    def perform(user_id:, reactable_id:, reactable_type:, category:)
+      user = User.find_by(id: user_id)
+      reactable = reactable_type.constantize.find_by(id: reactable_id)
```
I suggest adding a check before `constantize` to be sure that we constantize a valid class.
It could be a guard clause `return unless %w(Acticle Comment).include?(reactable_type)` or raising an exception if the `reactable_type` is not in thos list

johncurcio replied:
Would it be better to use [safe_constantize](https://apidock.com/rails/String/safe_constantize) instead, because it returns ``nil`` if no class is found? Something like:

```ruby
reactable = reactable_type.safe_constantize&.find_by(id: reactable_id)
```

I'm slightly concerned ``%w(Article Comment)`` might not be an extensive list with all reactable classes in the future (not sure I need to worry about that).

lightalloy replied:
It's not only about the class could not exist, but it could be another class (not a valid `reactable`) which could cause weird behavior. I'm not also a fan of using `&` when it's possible to avoid it (it's not a "confident" code if we are constantly checking for `nil`)
Imo explicit checking also works as documentation, that's why I suggested it. In the future, it would be required to add the classes to the list (if needed) and add the corresponding tests.

lightalloy commented on spec/jobs/reactions/create_job_spec.rb:
```diff
@@ -0,0 +1,50 @@
+      end
+    end
+
+    context "when user + reactable" do
```
Stubbing is a fine technique, but sometimes we can miss an error by using too much of it. I would add one test checking that the reaction is really created.

lightalloy commented on app/jobs/reactions/create_job.rb:
```diff
@@ -0,0 +1,19 @@
+
+      return unless user && reactable
+
+      Reaction.create(
```
`create` will fail silently if something is wrong. If you use `create!` instead of it, an exception will be raised and we'll notice the error.

johncurcio replied:
I did not know that! Thank you for catching this. I'm going to fix it.

johncurcio commented:
@lightalloy I changed the things you requested. If you have any other suggestions, let me know and I will happily apply them :)
